internal/decision/engine.go

```diff
@@ -167,16 +165,65 @@ type Engine struct {
 	sendDontHaves bool
 
 	self peer.ID
+
+	// metrics gauge for total pending tasks across all workers
+	pendingGauge metrics.Gauge
+
+	// metrics gauge for total pending tasks across all workers
+	activeGauge metrics.Gauge
+
+	// used to ensure metrics are reported each fixed number of operation
+	metricUpdateCounter int
 }
 
-// NewEngine creates a new block sending engine for the given block store
-func NewEngine(bs bstore.Blockstore, bstoreWorkerCount int, peerTagger PeerTagger, self peer.ID, scoreLedger ScoreLedger) *Engine {
-	return newEngine(bs, bstoreWorkerCount, peerTagger, self, maxBlockSizeReplaceHasWithBlock, scoreLedger)
+// NewEngine creates a new block sending engine for the given block store.
+// maxOutstandingBytesPerPeer hints to the peer task queue not to give a peer more tasks if it has some maximum
+// work already outstanding.
+func NewEngine(
+	ctx context.Context,
+	bs bstore.Blockstore,
+	bstoreWorkerCount,
+	engineTaskWorkerCount, maxOutstandingBytesPerPeer int,
+	peerTagger PeerTagger,
+	self peer.ID,
+	scoreLedger ScoreLedger,
+	pendingEngineGauge metrics.Gauge,
+	activeEngineGauge metrics.Gauge,
+	pendingBlocksGauge metrics.Gauge,
+	activeBlocksGauge metrics.Gauge,
```

Comment:
this is kind of annoying (as you saw in the tests) and will also cause breakages if we add more metrics (as you saw in the tests).

We could use functional options, but that's likely overkill. Maybe if you add a function to make modifying the tests to take new metrics less work than we can leave this as is since `NewEngine` is an internal function anyway.